File: services/server/src/safe_socket/safe_socket.py
import socket


def recv_all(socket: socket.socket, size):
    received = b""
    while len(received) < size:
        chunk = socket.recv(size - len(received))
        if not chunk:
            if received:
                # Peer closed the connection mid-message: we got some bytes but not all of them
                raise ConnectionError(
                    f"connection closed after receiving {len(received)} of {size} expected bytes"
                )
            # Peer closed the connection cleanly before sending anything new
            return b""
        received += chunk
    return received


def send_all(socket: socket.socket, bytes):
    sent = 0
    while sent < len(bytes):
        # send() may only accept part of the buffer per call (short write); a real closed
        # connection raises instead of returning 0, so just resume with the remaining bytes
        sent += socket.send(bytes[sent:])

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import (
	"fmt"
	"io"
)

func SendAll(socket io.Writer, bytes []byte) error {
	written := 0
	for written < len(bytes) {
		n, err := socket.Write(bytes[written:])
		written += n
		if err != nil {
			// Write can fail partway through; whatever it did manage to write is already counted above
			return err
		}
	}
	return nil
}

func RecvAll(socket io.Reader, size int) ([]byte, error) {
	buff := make([]byte, size)
	read := 0
	for read < size {
		n, err := socket.Read(buff[read:])
		read += n
		if err != nil {
			if err == io.EOF {
				// Peer closed the connection before sending all the bytes this call expected
				return nil, fmt.Errorf("connection closed after receiving %d of %d expected bytes", read, size)
			}
			return nil, err
		}
	}
	return buff, nil
}
